### PR TITLE
Improve TestedVersions to support comments, sorting

### PR DESCRIPTION
**Pull request by mmanciop**

### CODE REVIEW

File: noxfile.py [hunk 1]
```diff
@@ -34,10 +34,11 @@ def get_versions_from_pypi(package_name: str) -> List[str]:
     response = requests.get(f"https://pypi.org/rss/project/{package_name}/releases.xml")
     response.raise_for_status()
     xml_tree = ElementTree.fromstring(response.text)
-    versions = [i.text for i in xml_tree.findall("channel/item/title") if i.text]
-    # Verify all strings have the format for version
-    [Version(version) for version in versions]
```
Comment (nirLumigo) [inline noxfile.py:39]: Why did you delete it?

File: src/ci/tested_versions_utils.py [hunk 2]
```diff
@@ -1,38 +1,180 @@
...
+
+    def __eq__(self, other):
+        if isinstance(other, SemanticVersion):
+           return False
+
+        return self.version == other.version
+
+    def __lt__(self, other):
+        if isinstance(other, SemanticVersion):
+           return True
```
Comment (nirLumigo) [inline src/ci/tested_versions_utils.py:44]: Why?

File: src/ci/tested_versions_utils.py [hunk 3]
```diff
@@ -1,38 +1,180 @@
...
+class SemanticVersion:
+    supported: bool
+    major: int
+    minor: int
+    patch: int
+    suffix: str
+    comment: str
+
+    @property
+    def version(self):
```
Comment (nirLumigo) [inline src/ci/tested_versions_utils.py:59]: Please keep the original version string.

File: src/ci/tested_versions_utils.py [hunk 4]
```diff
@@ -1,38 +1,180 @@
...
+
+
+@dataclass(frozen=True)
+class NonSemanticVersion:
+    supported: bool
+    version: str
+    comment: str
+
+    def __eq__(self, other):
+        if isinstance(other, SemanticVersion):
```
Comment (nirLumigo) [inline src/ci/tested_versions_utils.py:37]:
```suggestion
        if not isinstance(other, NonSemanticVersion):
```

File: src/ci/tested_versions_utils.py [hunk 5]
```diff
@@ -1,38 +1,180 @@
...
+    patch: int
+    suffix: str
+    comment: str
+
+    @property
+    def version(self):
+        return f"{self.major}.{self.minor}.{self.patch}{self.suffix or ''}"
+
+    def __eq__(self, other):
+        if isinstance(other, NonSemanticVersion):
```
Comment (nirLumigo) [inline src/ci/tested_versions_utils.py:63]:
```suggestion
        if not isinstance(other, SemanticVersion):
```

File: src/ci/tested_versions_utils.py [hunk 6]
```diff
@@ -1,38 +1,180 @@
...
+        if self.minor > other.minor:
+            return False
+
+        if self.patch < other.patch:
+            return True
+
+        if self.patch > other.patch:
+            return False
+
+        if not self.suffix and self.suffix:
```
Comment (nirLumigo) [inline src/ci/tested_versions_utils.py:95]: This means `True and False` or `False and True`, it will always result in `False`.

File: src/ci/tested_versions_utils.py [hunk 7]
```diff
@@ -1,38 +1,180 @@
...
+            return True
+
+        if self.suffix and not self.suffix:
```
Comment (nirLumigo) [inline src/ci/tested_versions_utils.py:98]: This means `True and False` or `False and True`, it will always result in `False`.

File: src/ci/tested_versions_utils.py [hunk 8]
```diff
@@ -1,38 +1,180 @@
...
+            return False
+
+        return self.suffix < other.suffix
+
+
+def parse_version(version: str) -> Union[SemanticVersion, NonSemanticVersion]:
```
Comment (nirLumigo) [inline src/ci/tested_versions_utils.py:104]: Please run this function on all the existing versions to verify that it works correctly and doesn't crash on "strange" versions formats.

File: src/ci/tested_versions_utils.py [hunk 9]
```diff
@@ -1,38 +1,180 @@
...
 import os
+import re
 from contextlib import contextmanager
 from dataclasses import dataclass
-from typing import Tuple
+from typing import List, Union
+
+# Major, minor, patch and (non semver standard, suffix)
+_semanticVersionPattern = re.compile(r'(!)?(\d+).(\d+).(\d+)([^\s]*)(?:\s*#\s*(.*))?')
+_splitVersionFromCommentPattern = re.compile(r'(!)?([^\s]*)(?:\s*#\s*(.*))?')
```
Comment (nirLumigo) [inline src/ci/tested_versions_utils.py:11]: Can you change it to capital letters with underscores separating words? https://peps.python.org/pep-0008/#constants

Comment (nirLumigo) [inline src/ci/tested_versions_utils.py:11]: Can you add comments to explain the regex?

File: src/ci/tested_versions_utils.py [hunk 10]
```diff
@@ -1,38 +1,180 @@
...
+    res = re.search(_semanticVersionPattern, version)
+
+    if res:
+        (supported, major, minor, patch, suffix, comment) = res.groups()
+        # The `supported` is either an empty string (supported) or the '!' string (not supported)
+        return SemanticVersion(not bool(supported), int(major), int(minor), int(patch), suffix, comment)
+
+    (supported, version, comment) = re.search(_splitVersionFromCommentPattern, version).groups()
+    # The `supported` is either an empty string (supported) or the '!' string (not supported)
+    return NonSemanticVersion(not bool(supported), version, comment)
```
Comment (nirLumigo) [inline src/ci/tested_versions_utils.py:114]:
```suggestion
    return NonSemanticVersion(not supported, version, comment)
```
The `bool` is redundant.

File: src/ci/tested_versions_utils.py [hunk 11]
```diff
@@ -1,38 +1,180 @@
...
+        return self.suffix < other.suffix
+
+
+def parse_version(version: str) -> Union[SemanticVersion, NonSemanticVersion]:
+    res = re.search(_semanticVersionPattern, version)
+
+    if res:
+        (supported, major, minor, patch, suffix, comment) = res.groups()
+        # The `supported` is either an empty string (supported) or the '!' string (not supported)
+        return SemanticVersion(not bool(supported), int(major), int(minor), int(patch), suffix, comment)
```
Comment (nirLumigo) [inline src/ci/tested_versions_utils.py:110]:
```suggestion
        return SemanticVersion(not supported, int(major), int(minor), int(patch), suffix, comment)
```
The `bool` is redundant.

File: src/ci/tested_versions_utils.py [hunk 12]
```diff
@@ -1,38 +1,180 @@
...
+            if parsed_version.comment != previous_version.comment:
+                if parsed_version.comment:
+                    print(f"Updating comment for version '{parsed_version.version}' in {path}")
+                else:
+                    print(f"Removing comment for version '{parsed_version.version}' in {path}")
+        else:
+            print(f"Adding the {parsed_version.version} as {'supported' if parsed_version.supported else 'not supported'} to {path}")
+
+        with open(path, "w") as f:
+            for tested_version in sorted(tested_versions.versions):
```
Comment (nirLumigo) [inline src/ci/tested_versions_utils.py:161]: Please sort the versions files as part of this PR, otherwise the nightly will create bigger PRs which will be harder to understand.

File: src/ci/tested_versions_utils.py [hunk 13]
```diff
@@ -1,38 +1,180 @@
...
+#
+# ```
+# >>> from packaging.version import parse
+# >>> str(parse("1.2.4b"))
+# '1.2.4b0'
+# ```
+
+
+@dataclass(frozen=True)
+class NonSemanticVersion:
```
Comment (nirLumigo) [inline src/ci/tested_versions_utils.py:31]: Do we have such versions? I worry that maybe if a version will be identified as `NonSemanticVersion` it means that we have a bug at parsing the version. Maybe we should crash in this case instead of creating `NonSemanticVersion`?

File: scripts/gather_version_artifacts.py [hunk 14]
```diff
@@ -10,20 +10,12 @@
 # Before this script runs, the job downloads the artifacts into files with the following example structure:
 #
 #   versions_artifacts/
-#       3.7/
```
Comment (nirLumigo) [inline scripts/gather_version_artifacts.py:13]: Why did you delete it? The location of the artifacts wasn't changed.

**Review by nirLumigo — CHANGES_REQUESTED** [review]
Nice :)

File: scripts/gather_version_artifacts.py (hunk 14 above)
Comment (mmanciop) [reply scripts/gather_version_artifacts.py:13]: I misread the comment :-)

File: src/ci/tested_versions_utils.py (hunk 13 above)
Comment (mmanciop) [reply src/ci/tested_versions_utils.py:31]: I believe we will find non-semantic versions in NPM and PIPY

File: src/ci/tested_versions_utils.py [hunk 15]
```diff
@@ -1,38 +1,180 @@
...
+            else:
+                print(f"Turning version '{parsed_version.version}' to {'supported' if parsed_version.supported else 'not supported'} in {path}")
+
+            if parsed_version.comment != previous_version.comment:
+                if parsed_version.comment:
+                    print(f"Updating comment for version '{parsed_version.version}' in {path}")
+                else:
+                    print(f"Removing comment for version '{parsed_version.version}' in {path}")
+        else:
+            print(f"Adding the {parsed_version.version} as {'supported' if parsed_version.supported else 'not supported'} to {path}")
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/ci/tested_versions_utils.py:158]: But now the output does is not correct. It would show you `Replacing 1.2.3 with 1.2.3 in {path}` in all cases.

File: noxfile.py (hunk 1 above)
Comment (mmanciop) [reply noxfile.py:39]: Because that validation provides no value that I can speak of, and it will instead risk of breaking with non-Python packages. Remember, that `Version` object has some _strong_, PEP-related assumptions.

File: src/ci/tested_versions_utils.py (hunk 13 above)
Comment (mmanciop) [reply src/ci/tested_versions_utils.py:31]: Actually, we _already_ have non semver around: https://github.com/lumigo-io/opentelemetry-python-distro/blob/770255e8bec507d5e186[%E2%80%A6]_opentelemetry/instrumentations/pymysql/tested_versions/pymysql

File: requirements.txt [hunk 16]
```diff
@@ -17,3 +17,4 @@ zipp==3.8.0
 nox==2022.1.7
 psutil==5.9.1
 packaging==21.3
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply requirements.txt]: Done in 4bd946e8de5beb1cb30f23d2dc791460ffe562df

File: src/ci/tested_versions_utils.py (hunk 5 above)
Comment (mmanciop) [reply src/ci/tested_versions_utils.py]: Done in dfb0576a9213e84b59b2b3970f5c8738923abc7a

File: src/ci/tested_versions_utils.py (hunk 4 above)
Comment (mmanciop) [reply src/ci/tested_versions_utils.py]: Done in dfb0576a9213e84b59b2b3970f5c8738923abc7a

File: src/ci/tested_versions_utils.py (hunk 3 above)
Comment (mmanciop) [reply src/ci/tested_versions_utils.py]: Done in 148061654d9421eb40925ca8b23227f5fee565cd

File: src/ci/tested_versions_utils.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (nirLumigo) [reply src/ci/tested_versions_utils.py]: Then we can add the comment to the log as well.
I prefer to print all the data then selecting exactly what to print.

File: src/ci/tested_versions_utils.py (hunk 6 above)
Comment (mmanciop) [reply src/ci/tested_versions_utils.py]: Copy & paste issue, done in 91d68b568bee3e6a6daee826d0049e9437417612

File: src/ci/tested_versions_utils.py (hunk 7 above)
Comment (mmanciop) [reply src/ci/tested_versions_utils.py]: Copy & paste issue, done in 91d68b568bee3e6a6daee826d0049e9437417612

File: src/ci/tested_versions_utils.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/ci/tested_versions_utils.py]: Done in 91d68b568bee3e6a6daee826d0049e9437417612 through e9ab44feffd740d8cf637990984014d305bd25f3

**Review by nirLumigo — APPROVED** [review]

File: src/ci/tested_versions_utils.py [hunk 17]
```diff
@@ -151,7 +152,9 @@ def parse_version(version: str) -> Union[SemanticVersion, NonSemanticVersion]:
 
 @dataclass(frozen=True)
 class TestedVersions:
-    versions: List[Union[SemanticVersion, NonSemanticVersion]]
+    versions: List[Union[SemanticVersion, NonSemanticVersion]] = attrib(
```
Comment (nirLumigo) [inline src/ci/tested_versions_utils.py:155]: Cool! 💪

Comment (nirLumigo) [inline src/ci/tested_versions_utils.py:155]: Notice this warning from the documentation
image

File: src/test/ci/test_tested_versions.py [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+
+def test_parse_version():
+    assert parse_version("  1.2.3  # test!") == SemanticVersion(
+        True, "1.2.3  # test!", 1, 2, 3, "", "test!"
+    )
+
+    # We disregard comments in the equality check
+    assert parse_version("!1.2.3") == SemanticVersion(
+        False, "1.2.3  # test!", 1, 2, 3, "", "test!"
+    )
```
Comment (nirLumigo) [inline src/test/ci/test_tested_versions.py:17]: Usually we put each block like this in a different test and the comment inside is the name of the test. You can use `pytest.mark.parametrize` to make it nicer.
Example:
image

File: src/test/ci/test_tested_versions.py [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+            [
+                parse_version("!0.0.1"),
+                parse_version("0.0.2"),
+                parse_version("0.0.3"),
+                parse_version("0.1.0"),
+                parse_version("!0.1.1"),
+                parse_version("1.0.1"),
+            ]
+        )
+    ) == ["0.0.2~0.1.0", "1.0.1"]
```
Comment (nirLumigo) [inline src/test/ci/test_tested_versions.py:149]: Can you add a test with Non-semantic versions?

**Review by nirLumigo — CHANGES_REQUESTED** [review]
Nice! :)

File: src/test/ci/test_tested_versions.py (hunk 18 above)
Comment (mmanciop) [reply src/test/ci/test_tested_versions.py:17]: Yeah, but do you want to compare the understandably of the current status and using a parametrization to.... what? save you from typing `assert`? It may look cooler, but understandability goes down a hole so deep that you eventually find yourself in China.

File: src/test/ci/test_tested_versions.py (hunk 19 above)
Comment (mmanciop) [reply src/test/ci/test_tested_versions.py:150]: Done in 6b3f68136a98a941c421ca399cf9c8a38678d580

File: src/ci/tested_versions_utils.py (hunk 17 above)
Comment (mmanciop) [reply src/ci/tested_versions_utils.py:155]: good catch

Comment (mmanciop) [reply src/ci/tested_versions_utils.py]: Done in 730ed4f265e9fb3d097c8032be504507ee0bcbfd

Comment (mmanciop) [reply src/ci/tested_versions_utils.py]: Turns out `attr.s` has its own flavor of `frozen` that we can use instead of the `@dataclass`. And since we don't need sorting or equality checks or anything else for `TestedVersions`, we might as well jump over for this case.

**Review by nirLumigo — APPROVED** [review]

File: src/test/ci/test_tested_versions.py (hunk 18 above)
Comment (mmanciop) [reply src/test/ci/test_tested_versions.py]: Tests split in b4df02d864879cc357328295b6d332c45ba553b5
